addon/plugins/variable-plugin/variables/location.ts

```diff
@@ -0,0 +1,117 @@
+import {
+  DCT,
+  EXT,
+} from '@lblod/ember-rdfa-editor-lblod-plugins/utils/constants';
+import {
+  createEmberNodeSpec,
+  createEmberNodeView,
+  EmberNodeConfig,
+} from '@lblod/ember-rdfa-editor/utils/ember-node';
+import { v4 as uuidv4 } from 'uuid';
+import { DOMOutputSpec, PNode } from '@lblod/ember-rdfa-editor';
+import {
+  isVariable,
+  parseLabel,
+  parseVariableInstance,
+  parseVariableSource,
+  parseVariableType,
+} from '@lblod/ember-rdfa-editor-lblod-plugins/plugins/variable-plugin/utils/attribute-parsers';
+
+const CONTENT_SELECTOR = `span[property~='${EXT('content').prefixed}'],
+                          span[property~='${EXT('content').full}']`;
+
+const parseDOM = [
+  {
+    tag: 'span',
+    getAttrs: (node: HTMLElement) => {
+      if (
+        isVariable(node) &&
+        node.querySelector(CONTENT_SELECTOR) &&
+        parseVariableType(node) === 'location'
+      ) {
+        const mappingResource = node.getAttribute('resource');
+        if (!mappingResource) {
+          return false;
+        }
+        const variableInstance = parseVariableInstance(node);
+        const source = parseVariableSource(node);
+        const label = parseLabel(node);
+        return {
+          variableInstance:
+            variableInstance ?? `http://data.lblod.info/variables/${uuidv4()}`,
+          mappingResource,
+          source,
+          label,
+        };
+      }
+
+      return false;
+    },
+    contentElement: CONTENT_SELECTOR,
+  },
+];
+
+const toDOM = (node: PNode): DOMOutputSpec => {
+  const { mappingResource, variableInstance, source, label } = node.attrs;
+
+  const sourceSpan = source
+    ? [
+        [
+          'span',
+          {
+            property: DCT('source').prefixed,
+            resource: source as string,
+          },
+        ],
+      ]
+    : [];
+  return [
+    'span',
+    {
+      resource: mappingResource as string,
+      typeof: EXT('Mapping').prefixed,
+      'data-label': label as string,
+    },
+    [
+      'span',
+      { property: EXT('instance'), resource: variableInstance as string },
+    ],
+    ['span', { property: DCT('type').prefixed, content: 'location' }],
+    ...sourceSpan,
+    [
+      'span',
+      {
+        property: EXT('content').prefixed,
+      },
+      0,
+    ],
+  ];
+};
```

Comment:
there are some parse-helpers (in attribute-parsers.ts) which are really nice and make everything so much clearer. I do miss that it doesn't have its opposite in the same file to use in toDOM. It would be nice to also decrease the complexity of toDOM in some way, because the "core" is quite simple, but it looks very complex with all the lists (ugh).

I am not sure if you can easily make helpers that do what is needed though. But it feels worth it to maybe try :D

Reply:
Yes I know what you mean. I don't know if it can be improved in a much clearer way like I did in the parseDOM function though :(

Reply:
Could do something like
```
typeSpan(type) { return ['span', { property: DCT('type').prefixed, content: type }] }
contentSpan(content?) {return [
      'span',
      {
        property: EXT('content').prefixed,
      },
      content ? content : 0,
    ],
    
```

It feels a bit over-abstraction, but does make it easier to read `toDOM` and gives a duality between parseDOM and toDOM in functions. So idk :shrug: Would only make it clear if it is a good idea with bigger and harder to understand parse/toDOM combinations.

I don't know if that's better though and let you decide. I'm fine with leaving how it is too :)

Reply:
I think the main reason I'd find it easier to read is probably that we don't allow long lines, so even a simple span is over 4 lines now, while with the helpers it would be just one line and much shorter.
But that's abstracting because of our choice of formatting configuration. So I'm not gonna say that's a good reason to do so :p

Reply:
I added some helper functions, I think it should be more readable now :)